### ISSUE TITLE
`TensorIterator`: provide a two-argument version of `set_output`

### ISSUE THREAD

**Issue by nkaretnikov** [pytorch/pytorch#70242]
### 🚀 The feature, motivation and pitch

Working on porting an op to `TensorIterator`.

With a non-`TensorIterator` structured kernel, I'm able to use a two-argument version of `set_output` as described in the [structured kernel RFC](https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#functions-to-define).

I've tried doing the same in the `META` kernel that uses `TensorIterator` and I get an error saying that this two-argument variant is not found.

I think it happens because the method is looked up in the `TensorIterator` class first, which simply doesn't have this declaration. I think you can still reach the two-argument one by doing `MetaBase::set_output(input1.sizes(), input1.options());`, but it seems too verbose to me. So I'm suggesting to add a two-argument version.

It may not sound like a big deal, but it took me a while to figure this out (because there are so many moving parts). I initially assumed that one of the arguments to `set_output` was wrong. Then I started checking whether anything changed recently WRT to `set_output`. So not ideal, IMO.

Here's an example for a `META` kernel that uses `TensorIterator`:

```cpp
  // auto shape = at::infer_size(input1.sizes(), input2.sizes());  // was using this originally hence the confusion
  MetaBase::set_output(input1.sizes(), input1.options());  // this compiles
  // set_output(input1.sizes(), input1.options());  // this doesn't compile, I'm proposing to add this one
  set_output(0, input1.sizes(), {}, input1.options(), {}); // this compiles as well, but verbose
```

Relevant declarations:
- https://github.com/pytorch/pytorch/blob/70ed4f3ffc0af1e12f9da7f64104c9631a6d70ef/aten/src/ATen/TensorIterator.h#L575 (no two-argument version)
- https://github.com/pytorch/pytorch/blob/70ed4f3ffc0af1e12f9da7f64104c9631a6d70ef/aten/src/ATen/TensorMeta.h#L62

cc @Lezcano

**Comment by Lezcano** [comment]
So, the issue here is that, when doing structured kernels via TensorIterator, the structured kernel assumes that the output you gave it will be the output of the function. Alas, this is not true in general. For example, all loss functions have the structure of a componentwise operation and a reduction. That way, you do use TI, but you then apply a reduction on the output.

This could be solved by having `TensorIteratorBase` not calling `set_output` directly, but expecting this call to happen in the META function. Now, I am not 100% sure of what are the implications of this.

As it stands, I don't think structured kernels allow for implementing this pattern. What do you think @bdhirsh?

**Comment by bdhirsh** [comment]
Yeah, this came up recently with a port for `smooth_l1_loss`, which has the same pattern of using Tensoriterator to create an intermediate (and reducing it to get the output). We ended up porting it without using `TensorIteratorBase`, which isn't too bad (we pay the cost of one extra construction of TensorIterator in the impl function).

**Comment by Lezcano** [comment]
Isn't building TI twice fairly expensive for these operations? cc @jbschlosser
